### PR TITLE
Move object lifetime default computation

### PR DESCRIPTION
**Pull request by cjgillot**
Split from https://github.com/rust-lang/rust/pull/97393

r? @jackh726

### CODE REVIEW

File: compiler/rustc_typeck/src/collect/object_lifetime_defaults.rs [hunk 1]
```diff
@@ -0,0 +1,603 @@
...
+                ObjectLifetimeDefault::Empty => {
+                    if in_body {
+                        None
+                    } else {
+                        Some(Region::Static)
+                    }
+                }
+                ObjectLifetimeDefault::Static => Some(Region::Static),
+                ObjectLifetimeDefault::Param(def_id) => {
+                    let index = generics.param_def_id_to_index[&def_id];
```
Comment (cjgillot) [inline compiler/rustc_typeck/src/collect/object_lifetime_defaults.rs:496]: I'm not sure this is the right index here.

File: compiler/rustc_middle/src/query/mod.rs [hunk 2]
```diff
@@ -1594,6 +1594,11 @@ rustc_queries! {
     query object_lifetime_defaults(_: LocalDefId) -> Option<&'tcx [ObjectLifetimeDefault]> {
         desc { "looking up lifetime defaults for a region on an item" }
     }
+    /// Fetch the lifetimes for all the trait objects in an item-like.
+    query object_lifetime_map(_: LocalDefId) -> FxHashMap<ItemLocalId, Region> {
+        storage(ArenaCacheSelector<'tcx>)
+        desc { "looking up lifetime defaults for a region on an item" }
```
Comment (jackh726) [inline compiler/rustc_middle/src/query/mod.rs:1600]: Is this description correct?

It's probably worth elaborating a bit in a comment what the different between `object_lifetime_map` and `object_lifetime_defaults` are.

File: src/test/ui/issues/issue-41139.rs [hunk 3]
```diff
@@ -8,5 +8,5 @@ fn main() {
     // This isn't great. The issue here is that `dyn Trait` is not sized, so
     // `dyn Fn() -> dyn Trait` is not well-formed.
     let t: &dyn Trait = &get_function()();
-    //~^ ERROR expected function, found `&dyn Fn() -> (dyn Trait + 'static)`
+    //~^ ERROR expected function, found `&dyn Fn() -> dyn Trait`
```
Comment (jackh726) [inline src/test/ui/issues/issue-41139.rs:11]: Why did this change? Not a huge deal, but would be nice to know why

**Review by jackh726 — COMMENTED** [review]
Some initial comments.

File: compiler/rustc_typeck/src/collect.rs [hunk 4]
```diff
@@ -1343,8 +1348,13 @@ fn has_late_bound_regions<'tcx>(tcx: TyCtxt<'tcx>, node: Node<'tcx>) -> Option<S
             if self.has_late_bound_regions.is_some() {
                 return;
             }
+            if let hir::LifetimeName::ImplicitObjectLifetimeDefault = lt.name {
+                return;
+            }
```
[the remark this answers is not in the document]
Comment (cjgillot) [reply compiler/rustc_typeck/src/collect.rs:1353]: This code is only used for function item-likes.

I agree, this is slightly wrong.  However, this cannot cause a bug: `ImplicitObjectLifetimeDefault` can only reference a lifetime from an outer `&` reference, a generic argument, or `'static`.  So the lifetime this references has already been seen by the visitor.

However, this whole visitor will be useless once https://github.com/rust-lang/rust/pull/97720 lands.  All the lifetimes in a function's signature will be generic parameters, so only the loop in `has_late_bound_regions` will be required.

File: src/test/ui/issues/issue-41139.rs (hunk 3 above)
Comment (cjgillot) [reply src/test/ui/issues/issue-41139.rs:11]: This is a bug.

Comment (jackh726) [reply src/test/ui/issues/issue-41139.rs:11]: Wait, do you mean the existing test is a bug? Or the new output is?

Comment (cjgillot) [reply src/test/ui/issues/issue-41139.rs:11]: The new output is buggy.

**Comment by jackh726** [comment]
Gonna try to take another look at this PR this weekend.

**Review by jackh726 — APPROVED** [review]
Looking pretty good. Just two small comments. Also, can you merge some commits (pacify tidy, fix comments and query description) into other relevant commits. Not required, but would be nice. r=me after review addressed

**Comment by cjgillot** [comment]
I'm having doubts whether this is the right direction.  I'm marking this as blocked until I have a proper idea of where we are going.

**Pull request closed without merging**